account-compression/programs/account-compression/src/canopy.rs

```diff
@@ -128,3 +131,431 @@ pub fn fill_in_proof_from_canopy(
+        return Ok(()); // Canopy is empty
+    }
+    let mut empty_node_cache = Box::new([EMPTY; MAX_SUPPORTED_DEPTH]);
+    // first two nodes are the children of the root, they have index 2 and 3 respectively
+    let left_root_child =
+        get_value_for_node::<MAX_SUPPORTED_DEPTH>(2, max_depth - 1, canopy, &mut empty_node_cache);
+    let right_root_child =
+        get_value_for_node::<MAX_SUPPORTED_DEPTH>(3, max_depth - 1, canopy, &mut empty_node_cache);
+    let actual_root = hashv(&[&left_root_child, &right_root_child]).to_bytes();
+    if actual_root != *expected_root {
+        msg!(
+            "Canopy root mismatch. Expected: {:?}, Actual: {:?}",
+            expected_root,
+            actual_root
+        );
+        err!(AccountCompressionError::CanopyRootMismatch)
+    } else {
+        Ok(())
+    }
+}
+
+/// Checks the canopy doesn't have any nodes to the right of the provided index in the full tree.
+/// This is done by iterating through the canopy nodes to the right of the provided index and
+/// finding the top-most node that has the node as its left child. The node should be empty. The
+/// iteration contains following the previous checked node on the same level until the last node on
+/// the level is reached.
+pub fn check_canopy_no_nodes_to_right_of_index(
+    canopy_bytes: &[u8],
+    max_depth: u32,
+    index: u32,
+) -> Result<()> {
+    check_canopy_bytes(canopy_bytes)?;
+    check_index(index, max_depth)?;
+    let canopy = cast_slice::<u8, Node>(canopy_bytes);
+    let path_len = get_cached_path_length(canopy, max_depth)?;
+
+    let mut node_idx = ((1 << max_depth) + index) >> (max_depth - path_len);
+    // no need to check the node_idx as it's the leaf continaing the index underneath it
+    while node_idx & (node_idx + 1) != 0 {
+        // check the next node to the right
+        node_idx += 1;
+        // find the top-most node that has the node as its left-most child
+        node_idx >>= node_idx.trailing_zeros();
+
+        let shifted_index = node_idx as usize - 2;
+        if canopy[shifted_index] != EMPTY {
+            msg!("Canopy node at index {} is not empty", shifted_index);
+            return err!(AccountCompressionError::CanopyRightmostLeafMismatch);
+        }
+    }
+    Ok(())
+}
+
+#[inline(always)]
+fn check_index(index: u32, at_depth: u32) -> Result<()> {
+    if at_depth > MAX_SUPPORTED_DEPTH as u32 {
+        return err!(AccountCompressionError::ConcurrentMerkleTreeConstantsError);
+    }
+    if at_depth == 0 {
+        return err!(AccountCompressionError::ConcurrentMerkleTreeConstantsError);
+    }
+    if index >= (1 << at_depth) {
+        return err!(AccountCompressionError::LeafIndexOutOfBounds);
+    }
+    Ok(())
+}
+
+#[inline(always)]
+fn get_value_for_node<const N: usize>(
+    node_idx: usize,
+    level: u32,
+    canopy: &[Node],
+    empty_node_cache: &mut [Node; N],
+) -> Node {
+    if canopy[node_idx - 2] != EMPTY {
+        return canopy[node_idx - 2];
+    }
```

Comment:
If it wasn't for your extensive tests, I definitely would have thought this would be a footgun. 

Converting node indexing to canopy indexing almost always throws me off